README.md
# flutter_plot
R. C. Howell - 2018

A pretty plotting package for Flutter apps. Sizing and autopadding isn't great right now, but tinkering with padding and fontSize will allow for you to align things well. You can also check out `/docs` to learn more, but it really is a single file package.

## Example 1

![Screenshot](https://i.imgur.com/17QwWzg.png)

This example is more complicated than typical because it's showing off the styling capability.

```
final List<Point> data = [
  const Point(21.0, 19.0),
  const Point(3.0, 7.0),
  const Point(8.0, 9.0),
  const Point(11.0, 14.0),
  const Point(18.0, 17.0),
  const Point(7.0, 8.0),
  const Point(-4.0, -4.0),
  const Point(6.0, 12.0),
];

new Card(
  child: new Column(
    children: <Widget>[
      new Container(
        padding: const EdgeInsets.only(top: 12.0),
        child: new Text('Super Neat Plot'),
      ),
      new Container(
        child: new Plot(
          height: 200.0,
          data: widget.data,
          gridSize: new Offset(2.0, 2.0),
          style: new PlotStyle(
            pointRadius: 3.0,
            outlineRadius: 1.0,
            primary: Colors.white,
            secondary: Colors.orange,
            textStyle: new TextStyle(
              fontSize: 8.0,
              color: Colors.blueGrey,
            ),
            axis: Colors.blueGrey[600],
            gridline: Colors.blueGrey[100],
          ),
          padding: const EdgeInsets.fromLTRB(40.0, 12.0, 12.0, 40.0),
          xTitle: 'My X Title',
          yTitle: 'My Y Title',
          ),
      ),
    ],
  ),
),
```

## Example 2

![Screenshot 2](https://i.imgur.com/yPUG61p.png)

```
// Using the same data as before

Plot simplePlot = new Plot(
  height: 200.0,
  data: widget.data,
  gridSize: new Offset(2.0, 2.0),
  style: new PlotStyle(
    primary: Colors.black,
    textStyle: new TextStyle(
      fontSize: 8.0,
      color: Colors.blueGrey,
    ),
    axis: Colors.blueGrey[600],
  ),
  padding: const EdgeInsets.fromLTRB(40.0, 12.0, 12.0, 40.0),
);
```
## Displaying trace lines and coordinates

![Screenshot 4](https://i.imgur.com/DfEMb1Im.jpg)
![Screenshot 5](https://i.imgur.com/RkIYTycm.jpg)
![Screenshot 6](https://i.imgur.com/5M9h5zpm.jpg)

```
final List<Point> data = [
  const Point(9, 1),
  const Point(-7.0, 19.0),
  const Point(13.0, 5.0),
  const Point(-10.0, -4.0),
  const Point(16.0, 12.0),
];

new Card(
  child: new Column(
    children: <Widget>[
      new Container(
        padding: const EdgeInsets.only(top: 12.0),
        child: new Text('Super Neat Plot'),
      ),
      new Container(
        child: new Plot(
          height: 200.0,
          data: widget.data,
          gridSize: new Offset(2.0, 2.0),
          style: new PlotStyle(
            axisStrokeWidth: 2.0,
            pointRadius: 3.0,
            outlineRadius: 1.0,
            primary: Colors.yellow,
            secondary: Colors.red,
            trace: true,
            traceStokeWidth: 3.0,
            traceColor: Colors.blueGrey,
            traceClose: true,
            showCoordinates: true,
            textStyle: new TextStyle(
              fontSize: 8.0,
              color: Colors.grey,
            ),
            axis: Colors.blueGrey[600],
            gridline: Colors.blueGrey[100],
          ),
            axis: Colors.blueGrey[600],
            gridline: Colors.blueGrey[100],
          ),
          padding: const EdgeInsets.fromLTRB(40.0, 12.0, 12.0, 40.0),
          xTitle: 'My X Title',
          yTitle: 'My Y Title',
          ),
      ),
    ],
  ),
),
```

## How to Use
1. Add as a dependency
2. `import 'package:flutter_plot/flutter_plot.dart';`
3. See examples! There's not much to this package yet!


CHANGELOG.md
## [0.1.0]
Migrated to null-safety

## [0.0.3] - 4/18/2020
Updated PI constant to pi

## [0.0.2]

## [0.0.1] - 3/22/2018

See project README.md. This is the initial release of a small plotting library for Flutter.


test/flutter_plot_test.dart
import 'package:test/test.dart';

// import 'package:flutter_plot/plot.dart';

void main() {
  test('Tests a plot', () {

    // I honestly have no idea how one would go about testing a plotting tool programmatically, but I have
    // visually tested it many many times so take that for what you find it to be worth.

    // final plot = new Plot();
    // expect(a, b); expect a == b

  });
}


lib/plot.dart
library flutter_plot;

import 'package:flutter/material.dart';
import 'dart:ui';
import 'dart:math';
import 'package:meta/meta.dart';

/// Plot is a [StatelessWidget]
/// 
/// A [CustomPaint] is used to draw the Plot,
/// and the CustomPaint is wrapped in a [Container]
/// so that a user can provide height constraints and layout the
/// plot like they would other widgets. The plot takes all the
/// width that is available.
class Plot extends StatelessWidget {
  
  /// Widget height
  final double height;

  /// The dataset
  final List<Point> data;

  final PlotStyle style;

  /// gridSize determines how the grid is drawn
  /// 
  /// dx = x-width and dy = y-width.
  final Offset gridSize;

  /// Padding so that labels and text is visible
  /// 
  /// This is a known issue that someone more clever could help resolve.
  /// Ideally, padding should be calculated based upon the width/height of labels and 
  /// the xWindow/yWindow of the dataset, however, I'm unsure how to extract text width so
  /// that I may calculate padding.
  final EdgeInsets padding;

  /// Title to be displayed on the x-axis
  final String xTitle;

  /// Title to be displayed on the y-axis
  final String yTitle;

  Plot({
    this.height = 200.0,
    @required this.data,
    @required this.style,
    @required this.gridSize,
    @required this.padding,
    this.xTitle,
    this.yTitle,
  });

  @override
  Widget build(BuildContext context) {
    return new Container(
      height: this.height,
      padding: this.padding,
      alignment: Alignment.center,
      child: new CustomPaint(
        size: Size.infinite,
        painter: new _PlotPainter(
          points: this.data,
          style: this.style,
          gridSize: this.gridSize,
          xTitle: this.xTitle,
          yTitle: this.yTitle,
        ),
      ),
    );
  }

}

class _PlotPainter extends CustomPainter {

  final List<Point> points;
  final PlotStyle style;
  final Offset gridSize;
  final String xTitle;
  final String yTitle;

  // Used for determing the window
  double minX, maxX, minY, maxY, windowWidth, windowHeight;

  _PlotPainter({
    this.points,
    this.style,
    this.gridSize,
    this.xTitle,
    this.yTitle,
  }) : super() {
    assert (this.points != null && points.length > 0);
    this.points.sort((a, b) => a.x.compareTo(b.x));
    minX = (this.points.first.x > 0.0) ? 0.0 : this.points.first.x - gridSize.dx;
    maxX = (this.points.last.x < 0.0) ? 0.0 : this.points.last.x + gridSize.dx;
    this.points.sort((a, b) => a.y.compareTo(b.y));
    minY = (this.points.first.y > 0.0) ? 0.0 : this.points.first.y - gridSize.dy;
    maxY = (this.points.last.y < 0.0) ? 0.0 : this.points.last.y + gridSize.dy;
    windowWidth = maxX.abs() + minX.abs();
    windowHeight = maxY.abs() + minY.abs();
  }

  @override
  void paint(Canvas canvas, Size size) {

    Offset origin = _scalePoint(const Point(0.0, 0.0), size);

    // DRAWING GRIDLINES
    Paint gridPaint = new Paint();
    gridPaint.color = style.gridline ?? const Color(0x000000);
    gridPaint.style = PaintingStyle.stroke;
    gridPaint.strokeWidth = 1.0;

    int xGridlineCount = (windowWidth / gridSize.dx).round();
    int yGridlineCount = (windowHeight / gridSize.dy).round();
 
    for (int i = 0; i < xGridlineCount; i += 1) {
      // Routine for drawing xGridlines and xLabels
      drawLineAndLabel(double j) {
        Offset start = _scalePoint(new Point(j, 0), size);
        _drawXLabel((j).toString(), canvas, start.dx, origin.dy, size);
        if (style.gridline != null) {
          canvas.drawLine(
            new Offset(start.dx, 0.0),
            new Offset(start.dx, size.height),
            gridPaint
          );
        }
      }

      double x = i * gridSize.dx;
      if (-x >= minX && -x < 0) drawLineAndLabel(-x);
      if (x <= maxX && x >= 0) drawLineAndLabel(x);

    }

    for (int i = 0; i < yGridlineCount; i += 1) {
      // Routine for drawing yGridlines and yLabels
      drawLineAndLabel(double j) {
        Offset start = _scalePoint(new Point(minX, j), size);
        _drawYLabel((j).toString(), canvas, origin.dx, start.dy, size);
        if (style.gridline != null) {
          canvas.drawLine(
            new Offset(0.0, start.dy),
            new Offset(size.width, start.dy),
            gridPaint
          );
        }
      }
      double y = i * gridSize.dy;
      if (-y >= minY && -y < 0) drawLineAndLabel(-y);
      if (y <= maxY && y >= 0) drawLineAndLabel(y);
    }

    // DRAW xTitle
    if (xTitle != null) {
      TextPainter label = new TextPainter(
        text: new TextSpan(
          text: xTitle,
          style: this.style.textStyle ?? this.style.textStyle,
        ),
        textAlign: TextAlign.center,
        maxLines: 1,
        textDirection: TextDirection.ltr
      );
      label.layout(maxWidth: size.width, minWidth: size.width);
      Offset labelPos = new Offset(0.0, size.height + 2 * label.height);
      label.paint(canvas, labelPos);
    }

    // DRAW yTitle
    if (yTitle != null) {
      TextPainter label = new TextPainter(
        text: new TextSpan(
          text: yTitle,
          style: this.style.textStyle,
        ),
        textAlign: TextAlign.center,
        maxLines: 1,
        textDirection: TextDirection.ltr
      );
      label.layout(maxWidth: size.width, minWidth: size.width);
      Offset labelPos = new Offset(-1.5 * size.height, -3.5 * label.height);
      canvas.save();
      canvas.rotate(-PI / 2);
      label.paint(canvas, labelPos);
      canvas.restore();
    }

    // DRAWING EACH AXIS
    if (style.axis != null) {
      Paint axisPaint = new Paint();
      axisPaint.color = style.axis;
      axisPaint.style = PaintingStyle.stroke;
      axisPaint.strokeWidth = style.axisStrokeWidth;
      // Draw X Axis
      canvas.drawLine(
        new Offset(0.0, origin.dy),
        new Offset(size.width, origin.dy),
        axisPaint
      );
      // Draw Y Axis
      canvas.drawLine(
        new Offset(origin.dx, 0.0),
        new Offset(origin.dx, size.height),
        axisPaint
      );
    }

    // DRAWING THE POINTS
    Paint circlePaint = new Paint();
    circlePaint.color = style.primary;
    circlePaint.style = PaintingStyle.fill;
    Paint outlinePaint = new Paint();
    outlinePaint.color = style.secondary;
    outlinePaint.style = PaintingStyle.stroke;
    outlinePaint.strokeWidth = style.outlineRadius;
    this.points.forEach((Point p) {
      Offset point = _scalePoint(p, size);
      canvas.drawCircle(point, style.pointRadius, circlePaint);
      if (style.outlineRadius > 0.0) {
        canvas.drawCircle(point, style.pointRadius, outlinePaint);
      }
    });

  }

  bool shouldRepaint(_PlotPainter oldDelegate) => true;

  Offset _scalePoint(Point p, Size size) {
    double scaledX = (size.width * (p.x - minX)) / (maxX - minX);
    double scaledY = size.height - (size.height * (p.y - minY)) / (maxY - minY);
    return new Offset(scaledX, scaledY);
  }

  void _drawXLabel(String text, Canvas canvas, double x, double y, Size size) {
    TextPainter label = new TextPainter(
      text: new TextSpan(
        text: text,
        style: this.style.textStyle,
      ),
      textAlign: TextAlign.center,
      maxLines: 1,
      textDirection: TextDirection.ltr
    );
    label.layout(maxWidth: 20.0, minWidth: 10.0);
    Offset labelPos = new Offset(x - label.width / 2, size.height + label.height / 2);
    label.paint(canvas, labelPos);
  }

  void _drawYLabel(String text, Canvas canvas, double x, double y, Size size) {
    TextPainter label = new TextPainter(
      text: new TextSpan(
        text: text,
        style: this.style.textStyle,
      ),
      textAlign: TextAlign.right,
      maxLines: 1,
      textDirection: TextDirection.ltr
    );
    label.layout(maxWidth: 20.0, minWidth: 10.0);
    // Offset labelPos = new Offset(x-(label.width + 2.0), y + 3.0);
    Offset labelPos = new Offset(-label.width - 4.0, y - label.height / 2);
    label.paint(canvas, labelPos);
  }

}


/// PlotStyle is used to style a plot
class PlotStyle {

  /// Drawing radius of points
  /// 
  /// Defaults to 2.0
  final double pointRadius;

  /// Drawing radius of point outlines
  /// 
  /// Defaults to 0.0
  final double outlineRadius;

  /// Color to draw the points with
  /// 
  /// Defaults to #FF0000FF (Blue)
  final Color primary;

  /// Color to draw outlines with
  /// 
  /// Defaults to #FFFF0000 (Red)
  final Color secondary;

  /// Color to draw the gridlines
  /// 
  /// If not provided, then the gridlines are not drawn.
  final Color gridline;

  /// Color to draw the axis
  /// 
  /// If not provided, the axis is not drawn.
  final Color axis;

  /// Drawing width of the axis
  /// 
  /// Defaults to 1.0
  final double axisStrokeWidth;

  /// [TextStyle] for the axis labels and titles
  final TextStyle textStyle;

  PlotStyle({
    this.pointRadius = 2.0,
    this.outlineRadius = 0.0,
    this.primary = const Color(0xFF0000FF),
    this.secondary = const Color(0xFFFF0000),
    this.gridline,
    this.axis,
    this.axisStrokeWidth = 1.0,
    @required this.textStyle,
  });

}


lib/flutter_plot.dart
library flutter_plot;

import 'package:flutter/material.dart';
import 'dart:ui';
import 'dart:math';
import 'package:meta/meta.dart';

/// Plot is a [StatelessWidget]
///
/// A [CustomPaint] is used to draw the Plot,
/// and the CustomPaint is wrapped in a [Container]
/// so that a user can provide height constraints and layout the
/// plot like they would other widgets. The plot takes all the
/// width that is available.
class Plot extends StatelessWidget {
  /// Widget height
  final double height;

  /// The dataset
  final List<Point> data;

  final PlotStyle style;

  /// gridSize determines how the grid is drawn
  ///
  /// dx = x-width and dy = y-width.
  final Offset gridSize;

  /// Padding so that labels and text is visible
  ///
  /// This is a known issue that someone more clever could help resolve.
  /// Ideally, padding should be calculated based upon the width/height of labels and
  /// the xWindow/yWindow of the dataset, however, I'm unsure how to extract text width so
  /// that I may calculate padding.
  final EdgeInsets padding;

  /// Title to be displayed on the x-axis
  final String? xTitle;

  /// Title to be displayed on the y-axis
  final String? yTitle;

  /// If the origin should be centered
  final bool centered;

  Plot({
    this.height = 200.0,
    this.centered = false,
    required this.data,
    required this.style,
    required this.gridSize,
    required this.padding,
    this.xTitle,
    this.yTitle,
  });

  @override
  Widget build(BuildContext context) {
    return new Container(
      height: this.height,
      padding: this.padding,
      alignment: Alignment.center,
      child: new CustomPaint(
        size: Size.infinite,
        painter: new _PlotPainter(
            points: this.data,
            style: this.style,
            gridSize: this.gridSize,
            xTitle: this.xTitle,
            yTitle: this.yTitle,
            centered: this.centered),
      ),
    );
  }
}

class _PlotPainter extends CustomPainter {
  final List<Point> points;
  final PlotStyle style;
  final Offset gridSize;
  final String? xTitle;
  final String? yTitle;

  // Used for determing the window
  late double minX, maxX, minY, maxY, windowWidth, windowHeight;

  _PlotPainter(
      {required this.points,
      required this.style,
      required this.gridSize,
      this.xTitle,
      this.yTitle,
      bool centered = false})
      : super() {
    assert(points.length > 0);
    double minDataX =
        this.points.fold(0.0, (v, e) => v < e.x ? v : e.x.toDouble());
    minX = (minDataX > 0.0) ? 0.0 : minDataX - gridSize.dx;
    double maxDataX =
        this.points.fold(0, (v, e) => v < e.x ? e.x.toDouble() : v);
    maxX = (maxDataX < 0.0) ? 0.0 : maxDataX + gridSize.dx;
    double minDataY =
        this.points.fold(0.0, (v, e) => v < e.y ? v : e.y.toDouble());
    minY = (minDataY > 0.0) ? 0.0 : minDataY - gridSize.dy;
    double maxDataY =
        this.points.fold(0.0, (v, e) => v < e.y ? e.y.toDouble() : v);
    maxY = (maxDataY < 0.0) ? 0.0 : maxDataY + gridSize.dy;
    windowWidth = maxX.abs() + minX.abs();
    windowHeight = maxY.abs() + minY.abs();

    if (centered) {
      final double longestDistance =
          max(min(minX, minY).abs(), max(maxX, maxY));
      maxY = longestDistance;
      maxX = longestDistance;
      minY = longestDistance * -1;
      minX = longestDistance * -1;
      windowHeight = longestDistance * 2;
      windowWidth = longestDistance * 2;
    }
  }

  @override
  void paint(Canvas canvas, Size size) {
    Offset origin = _scalePoint(const Point(0.0, 0.0), size);

    // DRAWING GRIDLINES
    Paint gridPaint = new Paint();
    gridPaint.color = style.gridline ?? const Color(0x000000);
    gridPaint.style = PaintingStyle.stroke;
    gridPaint.strokeWidth = 1.0;

    int xGridlineCount = (windowWidth / gridSize.dx).round();
    int yGridlineCount = (windowHeight / gridSize.dy).round();

    for (int i = 0; i < xGridlineCount; i += 1) {
      // Routine for drawing xGridlines and xLabels
      drawLineAndLabel(num j) {
        Offset start = _scalePoint(new Point(j, 0), size);
        _drawXLabel((j).toString(), canvas, start.dx, origin.dy, size);
        if (style.gridline != null) {
          canvas.drawLine(new Offset(start.dx, 0.0),
              new Offset(start.dx, size.height), gridPaint);
        }
      }

      num x = i * gridSize.dx;
      if (!style.trailingZeros && x % 1 == 0) x = x.toInt();
      if (-x >= minX && -x < 0) drawLineAndLabel(-x);
      if (x <= maxX && x >= 0) drawLineAndLabel(x);
    }

    for (int i = 0; i < yGridlineCount; i += 1) {
      // Routine for drawing yGridlines and yLabels
      drawLineAndLabel(num j) {
        Offset start = _scalePoint(new Point(minX, j), size);
        _drawYLabel((j).toString(), canvas, origin.dx, start.dy, size);
        if (style.gridline != null) {
          canvas.drawLine(new Offset(0.0, start.dy),
              new Offset(size.width, start.dy), gridPaint);
        }
      }

      num y = i * gridSize.dy;
      if (!style.trailingZeros && y % 1 == 0) y = y.toInt();

      if (-y >= minY && -y < 0) drawLineAndLabel(-y);
      if (y <= maxY && y >= 0) drawLineAndLabel(y);
    }

    // DRAW xTitle
    if (xTitle != null) {
      TextPainter label = new TextPainter(
          text: new TextSpan(
            text: xTitle,
            style: this.style.textStyle ?? this.style.textStyle,
          ),
          textAlign: TextAlign.center,
          maxLines: 1,
          textDirection: TextDirection.ltr);
      label.layout(maxWidth: size.width, minWidth: size.width);
      Offset labelPos = new Offset(0.0, size.height + 2 * label.height);
      label.paint(canvas, labelPos);
    }

    // DRAW yTitle
    if (yTitle != null) {
      TextPainter label = new TextPainter(
          text: new TextSpan(
            text: yTitle,
            style: this.style.textStyle,
          ),
          textAlign: TextAlign.center,
          maxLines: 1,
          textDirection: TextDirection.ltr);
      label.layout(maxWidth: size.width, minWidth: size.width);
      Offset labelPos = new Offset(-1.5 * size.height, -3.5 * label.height);
      canvas.save();
      canvas.rotate(-pi / 2);
      label.paint(canvas, labelPos);
      canvas.restore();
    }

    // DRAWING EACH AXIS
    if (style.axis != null) {
      Paint axisPaint = new Paint();
      axisPaint.color = style.axis!;
      axisPaint.style = PaintingStyle.stroke;
      axisPaint.strokeWidth = style.axisStrokeWidth;
      // Draw X Axis
      canvas.drawLine(new Offset(0.0, origin.dy),
          new Offset(size.width, origin.dy), axisPaint);
      // Draw Y Axis
      canvas.drawLine(new Offset(origin.dx, 0.0),
          new Offset(origin.dx, size.height), axisPaint);
    }

    // DRAWING TRACE LINES
    if (style.trace) {
      Paint traceLinePaint = new Paint();
      traceLinePaint.color = style.traceColor ?? style.secondary;
      traceLinePaint.strokeWidth = style.traceStokeWidth;
      traceLinePaint.style = PaintingStyle.fill;
      for (int i = 0; i < this.points.length; i++) {
        bool isLastPoint = (i + 1) == this.points.length;
        var firstPoint = _scalePoint(this.points[i], size);
        if (!isLastPoint) {
          var secondPoint = _scalePoint(this.points[i + 1], size);
          canvas.drawLine(
            firstPoint,
            secondPoint,
            traceLinePaint,
          );
        }
        if (isLastPoint && style.traceClose && this.points.length > 2) {
          var secondPoint = _scalePoint(this.points[0], size);
          canvas.drawLine(
            firstPoint,
            secondPoint,
            traceLinePaint,
          );
        }
      }
    }

    // DRAWING THE POINTS
    Paint circlePaint = new Paint();
    circlePaint.color = style.primary;
    circlePaint.style = PaintingStyle.fill;
    Paint outlinePaint = new Paint();
    outlinePaint.color = style.secondary;
    outlinePaint.style = PaintingStyle.stroke;
    outlinePaint.strokeWidth = style.outlineRadius;
    this.points.forEach((Point p) {
      Offset point = _scalePoint(p, size);
      canvas.drawCircle(point, style.pointRadius, circlePaint);
      if (style.outlineRadius > 0.0) {
        canvas.drawCircle(point, style.pointRadius, outlinePaint);
      }

      // DRAWING POINT COORDINATES
      if (style.showCoordinates) {
        TextPainter coordlabel = new TextPainter(
            text: new TextSpan(
              text: "(${p.x}, ${p.y})",
              style: new TextStyle(
                color: Colors.white,
                fontSize: 10.0,
                fontWeight: FontWeight.bold,
              ),
            ),
            textAlign: TextAlign.center,
            maxLines: 1,
            textDirection: TextDirection.ltr);
        coordlabel.layout(maxWidth: 192.0, minWidth: 24.0);
        Offset labelPos = new Offset(point.dx - coordlabel.width / 2,
            point.dy - (coordlabel.height / 2) - 16.0);
        Offset backLabelTL =
            new Offset(point.dx - (coordlabel.width / 2) - 6, point.dy - 6);
        Offset backLabelBR =
            new Offset(point.dx + (coordlabel.width / 2) + 6, point.dy - 26.0);
        Rect rect = new Rect.fromPoints(backLabelTL, backLabelBR);
        RRect rRect = new RRect.fromRectAndRadius(rect, Radius.circular(10.0));
        Paint backLabelPaint = new Paint();
        backLabelPaint.color = Colors.grey;
        backLabelPaint.style = PaintingStyle.fill;
        canvas.drawRRect(rRect, backLabelPaint);
        coordlabel.paint(canvas, labelPos);
      }
    });
  }

  bool shouldRepaint(_PlotPainter oldDelegate) => true;

  Offset _scalePoint(Point p, Size size) {
    double scaledX = (size.width * (p.x - minX)) / (maxX - minX);
    double scaledY = size.height - (size.height * (p.y - minY)) / (maxY - minY);
    return new Offset(scaledX, scaledY);
  }

  void _drawXLabel(String text, Canvas canvas, double x, double y, Size size) {
    TextPainter label = new TextPainter(
        text: new TextSpan(
          text: text,
          style: this.style.textStyle,
        ),
        textAlign: TextAlign.center,
        maxLines: 1,
        textDirection: TextDirection.ltr);
    label.layout(maxWidth: 20.0, minWidth: 10.0);
    Offset labelPos =
        new Offset(x - label.width / 2, size.height + label.height / 2);
    label.paint(canvas, labelPos);
  }

  void _drawYLabel(String text, Canvas canvas, double x, double y, Size size) {
    TextPainter label = new TextPainter(
        text: new TextSpan(
          text: text,
          style: this.style.textStyle,
        ),
        textAlign: TextAlign.right,
        maxLines: 1,
        textDirection: TextDirection.ltr);
    label.layout(maxWidth: 20.0, minWidth: 10.0);
    // Offset labelPos = new Offset(x-(label.width + 2.0), y + 3.0);
    Offset labelPos = new Offset(-label.width - 4.0, y - label.height / 2);
    label.paint(canvas, labelPos);
  }
}

/// PlotStyle is used to style a plot
class PlotStyle {
  /// Drawing radius of points
  ///
  /// Defaults to 2.0
  final double pointRadius;

  /// Drawing radius of point outlines
  ///
  /// Defaults to 0.0
  final double outlineRadius;

  /// Color to draw the points with
  ///
  /// Defaults to #FF0000FF (Blue)
  final Color primary;

  /// Color to draw outlines with
  ///
  /// Defaults to #FFFF0000 (Red)
  final Color secondary;

  /// Color to draw the gridlines
  ///
  /// If not provided, then the gridlines are not drawn.
  final Color? gridline;

  /// Color to draw the axis
  ///
  /// If not provided, the axis is not drawn.
  final Color? axis;

  /// Drawing width of the axis
  ///
  /// Defaults to 1.0
  final double axisStrokeWidth;

  /// [TextStyle] for the axis labels and titles
  final TextStyle? textStyle;

  /// If true lines will be drawn between each consecutive point
  ///
  /// Defaluts to false
  bool trace;

  /// If true a line will be traced between the last point and the first
  ///
  /// Defaluts to false
  bool traceClose;

  /// Color to draw trace lines with
  ///
  /// Defaults to #FF0000FF (Blue)
  final Color? traceColor;

  /// Trace line stroke width
  ///
  /// Defaults to 2.0
  final double traceStokeWidth;

  /// If true each point's coordinates will be displayed above each point
  ///
  /// Defaults to false
  bool showCoordinates;

  /// If trailing zeros should be drawn
  ///
  /// Defaults to true
  final bool trailingZeros;

  PlotStyle({
    this.pointRadius = 2.0,
    this.outlineRadius = 0.0,
    this.primary = const Color(0xFF0000FF),
    this.secondary = const Color(0xFFFF0000),
    this.gridline,
    this.axis,
    this.axisStrokeWidth = 1.0,
    this.trace = false,
    this.traceClose = false,
    this.traceColor,
    this.traceStokeWidth = 2.0,
    this.showCoordinates = false,
    this.trailingZeros = true,
    required this.textStyle,
  });
}


example/lib/main.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:flutter_plot/flutter_plot.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final List<Point> data = [
    const Point(21.0, 19.0),
    const Point(3.0, 7.0),
    const Point(8.0, 9.0),
    const Point(11.0, 14.0),
    const Point(18.0, 17.0),
    const Point(7.0, 8.0),
    const Point(-4.0, -4.0),
    const Point(6.0, 12.0),
  ];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Flutter plot'),
      ),
      body: ListView(
        children: <Widget>[
          Card(
            child: new Column(
              children: <Widget>[
                new Container(
                  padding: const EdgeInsets.only(top: 12.0),
                  child: new Text('Super Neat Plot'),
                ),
                new Container(
                  child: new Plot(
                    height: 200.0,
                    data: data,
                    gridSize: new Offset(2.0, 2.0),
                    style: new PlotStyle(
                      pointRadius: 3.0,
                      outlineRadius: 1.0,
                      primary: Colors.white,
                      secondary: Colors.orange,
                      textStyle: new TextStyle(
                        fontSize: 8.0,
                        color: Colors.blueGrey,
                      ),
                      axis: Colors.blueGrey[600],
                      gridline: Colors.blueGrey[100],
                    ),
                    padding: const EdgeInsets.fromLTRB(40.0, 12.0, 12.0, 40.0),
                    xTitle: 'My X Title',
                    yTitle: 'My Y Title',
                  ),
                ),
              ],
            ),
          ),
          Card(
            child: Column(
              children: <Widget>[
                new Container(
                  padding: const EdgeInsets.only(top: 12.0),
                  child: new Text('Super Neat Plot'),
                ),
                new Container(
                  child: new Plot(
                    height: 200.0,
                    data: data,
                    gridSize: new Offset(2.0, 2.0),
                    style: new PlotStyle(
                      axisStrokeWidth: 2.0,
                      pointRadius: 3.0,
                      outlineRadius: 1.0,
                      primary: Colors.yellow,
                      secondary: Colors.red,
                      trace: true,
                      traceStokeWidth: 3.0,
                      traceColor: Colors.blueGrey,
                      traceClose: true,
                      showCoordinates: true,
                      textStyle: new TextStyle(
                        fontSize: 8.0,
                        color: Colors.grey,
                      ),
                      axis: Colors.blueGrey[600],
                      gridline: Colors.blueGrey[100],
                    ),
                    padding: const EdgeInsets.fromLTRB(40.0, 12.0, 12.0, 40.0),
                    //xTitle: 'My X Title',
                    //yTitle: 'My Y Title',
                  ),
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }
}
